pkg/kubectl/rolling_updater.go

```diff
@@ -93,27 +79,433 @@ const (
+	}
+
+	// This is the maximum number of replicas we'll scale in either direction.
+	maxIncrement := 1
+	// This is the minimum number of replicas (total of old/new) which must
+	// remain available for the duration of the update if progress is to be made.
+	minAvailable := oldRc.Spec.Replicas - maxIncrement
+	// This drives the scaling direction. The default is scale up/down. If
+	// false, use down/up (in-place).
+	scaleUpFirst := true
+	// By default, use an availability check which is based on pod existence to
+	// avoid waiting for readiness to make progress.
+	availabilityCheck := createdCheck
+	if config.UpdatePercent != nil {
+		percent := *config.UpdatePercent
+		// Compute the scale increment based on a percentage of the new desired
+		// count.
+		maxIncrement = int(math.Ceil(float64(desired) * (math.Abs(float64(percent)) / 100)))
+		// Compute the minimum available requirement as a percentage of the old
+		// replica count.
+		percentOld := int(math.Ceil(float64(oldRc.Spec.Replicas) * (math.Abs(float64(percent)) / 100)))
+		minAvailable = oldRc.Spec.Replicas - percentOld
+		// A negative percentage indicates our scale direction should be
+		// down/up instead of the default up/down.
+		if percent < 0 {
+			scaleUpFirst = false
+		}
+		// When using percentages, switch to a readiness based availability check.
+		availabilityCheck = readinessCheck
+	}
+	// Impose a floor of 1 for the minimum available unless this is a 100%
+	// down/up scaling operation. A down/up can't have this floor because it
+	// would be impossible to do an in-place update using a minimum available of
+	// 1 with a maximum overall pod count of 2 when scaling down first.
+	if config.UpdatePercent == nil || (*config.UpdatePercent > 0) {
+		minAvailable = int(math.Max(float64(1), float64(minAvailable)))
+	}
+	// Helpful output about what we're about to do.
+	direction := "down"
+	if scaleUpFirst {
+		direction = "up"
+	}
+	fmt.Fprintf(out, "Scaling up %s from %d to %d, scaling down %s from %d to 0 (scale %s first by %d each interval, maintain at least %d available)\n",
+		newRc.Name, newRc.Spec.Replicas, desired, oldRc.Name, oldRc.Spec.Replicas, direction, maxIncrement, minAvailable)
+
+	// Scale newRc and oldRc until newRc has the desired number of replicas and
+	// oldRc has 0 replicas.
+	increment := maxIncrement
+	for newRc.Spec.Replicas != desired || oldRc.Spec.Replicas != 0 {
+		if scaleUpFirst {
+			// Scale up/down. Initially this means scale up by the max increment.
+			scaledRc, err := r.scaleUp(newRc, oldRc, desired, increment, config)
+			if err != nil {
+				return err
+			}
+			newRc = scaledRc
+			scaleUpFirst = false
+		} // Otherwise, scale down/up.
+
+		// Scale down as much as possible while maintaining the minimum available
+		// amount. The scale-down informs us as to how much we can safely scale up
+		// next interval.
+		scaledRc, newIncrement, err := r.scaleDown(newRc, oldRc, desired, maxIncrement, minAvailable, availabilityCheck, config)
```

Comment:
Late in the day last week I think I remember detecting this with a unit test. I agree that we should handle the timeout here.